message.go

```diff
@@ -513,14 +513,78 @@ type Envelope struct {
 	MessageId string
 }
 
+// Permutations of the layouts defined in RFC 5322, section 3.3.
+var dateLayouts = [...]string{
+	MessageDateTimeLayout, // popular, try it first
+	"2 Jan 2006 15:04:05 -0700",
+	"2 Jan 2006 15:04:05 MST",
+	"2 Jan 2006 15:04:05 -0700 (MST)",
+	"2 Jan 2006 15:04 -0700",
+	"2 Jan 2006 15:04 MST",
+	"2 Jan 2006 15:04 -0700 (MST)",
+	"2 Jan 06 15:04:05 -0700",
+	"2 Jan 06 15:04:05 MST",
+	"2 Jan 06 15:04:05 -0700 (MST)",
+	"2 Jan 06 15:04 -0700",
+	"2 Jan 06 15:04 MST",
+	"2 Jan 06 15:04 -0700 (MST)",
+	"02 Jan 2006 15:04:05 -0700",
+	"02 Jan 2006 15:04:05 MST",
+	"02 Jan 2006 15:04:05 -0700 (MST)",
+	"02 Jan 2006 15:04 -0700",
+	"02 Jan 2006 15:04 MST",
+	"02 Jan 2006 15:04 -0700 (MST)",
+	"02 Jan 06 15:04:05 -0700",
+	"02 Jan 06 15:04:05 MST",
+	"02 Jan 06 15:04:05 -0700 (MST)",
+	"02 Jan 06 15:04 -0700",
+	"02 Jan 06 15:04 MST",
+	"02 Jan 06 15:04 -0700 (MST)",
+	"Mon, 2 Jan 2006 15:04:05 -0700",
+	"Mon, 2 Jan 2006 15:04:05 MST",
+	"Mon, 2 Jan 2006 15:04:05 -0700 (MST)",
+	"Mon, 2 Jan 2006 15:04 -0700",
+	"Mon, 2 Jan 2006 15:04 MST",
+	"Mon, 2 Jan 2006 15:04 -0700 (MST)",
+	"Mon, 2 Jan 06 15:04:05 -0700",
+	"Mon, 2 Jan 06 15:04:05 MST",
+	"Mon, 2 Jan 06 15:04:05 -0700 (MST)",
+	"Mon, 2 Jan 06 15:04 -0700",
+	"Mon, 2 Jan 06 15:04 MST",
+	"Mon, 2 Jan 06 15:04 -0700 (MST)",
+	"Mon, 02 Jan 2006 15:04:05 MST",
+	"Mon, 02 Jan 2006 15:04:05 -0700 (MST)",
+	"Mon, 02 Jan 2006 15:04 -0700",
+	"Mon, 02 Jan 2006 15:04 MST",
+	"Mon, 02 Jan 2006 15:04 -0700 (MST)",
+	"Mon, 02 Jan 06 15:04:05 -0700",
+	"Mon, 02 Jan 06 15:04:05 MST",
+	"Mon, 02 Jan 06 15:04:05 -0700 (MST)",
+	"Mon, 02 Jan 06 15:04 -0700",
+	"Mon, 02 Jan 06 15:04 MST",
+	"Mon, 02 Jan 06 15:04 -0700 (MST)",
+}
+
+// Try parsing the date based on the layouts defined in RFC 5322, section 3.3.
+// Inspired by https://github.com/golang/go/blob/master/src/net/mail/message.go
+func (e *Envelope) ParseDate(maybeDate string) (time.Time, error) {
```

Comment:
Yeah, that would be better. Also, I'm not sure we want to export this parse function as it's not directly related to IMAP (it should be defined in `net/mail` in a perfect world).

I can help if you want (I just don't want to do something you don't want me to :P ).